File: typespec/admin/users/management.ts
import type { EmailAddress } from "../../common/common.ts";
import type { DisplayName } from "../../common/localization.ts";
import type { PageSize, PaginationKey } from "../../common/pagination.ts";
import { isPageSize, isPaginationKey } from "../../common/pagination.ts";
import type {
  AdminAuthorization,
  AdminPermissionID,
} from "../authorization/types.ts";
import { validatePermissions } from "../authorization/types.ts";
import type { AdminUserID } from "../types.ts";
import { isAdminUserID } from "../types.ts";
import type { AdminUserState } from "./state.ts";

export type AdminUserFilterText = string;
export type AdminLastLoginFilter =
  | "never"
  | "inactive_30_days"
  | "inactive_90_days";

export interface ListUsersRequest {
  limit?: PageSize;
  pagination_key?: PaginationKey;
  filter_search?: AdminUserFilterText;
  filter_state?: AdminUserState;
  filter_permissions?: AdminPermissionID[];
  filter_no_permissions?: boolean;
  filter_totp_enabled?: boolean;
  filter_last_login?: AdminLastLoginFilter;
}

export function effectiveListUsersLimit(request: ListUsersRequest): PageSize {
  return request.limit ?? 50;
}

export function validateListUsersRequest(request: ListUsersRequest): string[] {
  const fields: string[] = [];
  if (!isPageSize(effectiveListUsersLimit(request))) {
    fields.push("limit");
  }
  if (
    request.pagination_key !== undefined &&
    !isPaginationKey(request.pagination_key)
  ) {
    fields.push("pagination_key");
  }
  if (
    request.filter_search !== undefined &&
    !isAdminUserFilterText(request.filter_search)
  ) {
    fields.push("filter_search");
  }
  if (
    request.filter_state !== undefined &&
    request.filter_state !== "active" &&
    request.filter_state !== "disabled"
  ) {
    fields.push("filter_state");
  }
  if (
    request.filter_permissions !== undefined &&
    !validatePermissions(request.filter_permissions)
  ) {
    fields.push("filter_permissions");
  }
  if (
    request.filter_last_login !== undefined &&
    request.filter_last_login !== "never" &&
    request.filter_last_login !== "inactive_30_days" &&
    request.filter_last_login !== "inactive_90_days"
  ) {
    fields.push("filter_last_login");
  }
  return fields;
}

function isAdminUserFilterText(value: AdminUserFilterText): boolean {
  const length = [...value].length;
  return length >= 1 && length <= 320;
}

export interface AdminUserSummary extends AdminAuthorization {
  admin_user_id: AdminUserID;
  email_address: EmailAddress;
  display_name: DisplayName;
  state: AdminUserState;
  totp_enabled: boolean;
  last_login_at?: string;
  created_at: string;
}

export interface ListUsersResponse {
  users: AdminUserSummary[];
  next_pagination_key?: PaginationKey;
}

export interface DisableUserRequest {
  admin_user_id: AdminUserID;
}

export function validateDisableUserRequest(
  request: DisableUserRequest,
): string[] {
  return isAdminUserID(request.admin_user_id) ? [] : ["admin_user_id"];
}

export interface EnableUserRequest {
  admin_user_id: AdminUserID;
}

export function validateEnableUserRequest(
  request: EnableUserRequest,
): string[] {
  return isAdminUserID(request.admin_user_id) ? [] : ["admin_user_id"];
}

File: typespec/admin/users/management.go
package users

import (
	"time"

	adminspec "github.com/vetchium/src/typespec/admin"
	"github.com/vetchium/src/typespec/admin/authorization"
	"github.com/vetchium/src/typespec/admin/user"
	"github.com/vetchium/src/typespec/common"
)

type AdminUserFilterText string
type AdminLastLoginFilter string

const (
	LastLoginNever          AdminLastLoginFilter = "never"
	LastLoginInactive30Days AdminLastLoginFilter = "inactive_30_days"
	LastLoginInactive90Days AdminLastLoginFilter = "inactive_90_days"
)

type ListUsersRequest struct {
	Limit               *common.PageSize                  `json:"limit,omitempty"`
	PaginationKey       *common.PaginationKey             `json:"pagination_key,omitempty"`
	FilterSearch        *AdminUserFilterText              `json:"filter_search,omitempty"`
	FilterState         *user.State                       `json:"filter_state,omitempty"`
	FilterPermissions   []authorization.AdminPermissionID `json:"filter_permissions,omitempty"`
	FilterNoPermissions *bool                             `json:"filter_no_permissions,omitempty"`
	FilterTOTPEnabled   *bool                             `json:"filter_totp_enabled,omitempty"`
	FilterLastLogin     *AdminLastLoginFilter             `json:"filter_last_login,omitempty"`
}

func (r ListUsersRequest) EffectiveLimit() common.PageSize {
	if r.Limit == nil {
		return 50
	}
	return *r.Limit
}

func (r ListUsersRequest) Validate() []string {
	fields := make([]string, 0, 7)
	if !common.IsPageSize(r.EffectiveLimit()) {
		fields = append(fields, "limit")
	}
	if r.PaginationKey != nil &&
		!common.IsPaginationKey(*r.PaginationKey) {
		fields = append(fields, "pagination_key")
	}
	if r.FilterSearch != nil && !isAdminUserFilterText(*r.FilterSearch) {
		fields = append(fields, "filter_search")
	}
	if r.FilterState != nil &&
		*r.FilterState != user.Active && *r.FilterState != user.Disabled {
		fields = append(fields, "filter_state")
	}
	if !authorization.ValidatePermissions(r.FilterPermissions) {
		fields = append(fields, "filter_permissions")
	}
	if r.FilterLastLogin != nil &&
		!isAdminLastLoginFilter(*r.FilterLastLogin) {
		fields = append(fields, "filter_last_login")
	}
	return fields
}

func isAdminLastLoginFilter(value AdminLastLoginFilter) bool {
	return value == LastLoginNever || value == LastLoginInactive30Days ||
		value == LastLoginInactive90Days
}

func isAdminUserFilterText(value AdminUserFilterText) bool {
	length := len([]rune(value))
	return length >= 1 && length <= 320
}

type AdminUserSummary struct {
	AdminUserID                adminspec.AdminUserID         `json:"admin_user_id"`
	EmailAddress               common.EmailAddress           `json:"email_address"`
	DisplayNames               []common.LocalizedDisplayName `json:"display_names"`
	PrimaryDisplayNameLanguage common.RegionalLanguageCode   `json:"primary_display_name_language"`
	State                      user.State                    `json:"state"`
	authorization.AdminAuthorization
	TOTPEnabled bool       `json:"totp_enabled"`
	LastLoginAt *time.Time `json:"last_login_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

type ListUsersResponse struct {
	Users             []AdminUserSummary    `json:"users"`
	NextPaginationKey *common.PaginationKey `json:"next_pagination_key,omitempty"`
}

type DisableUserRequest struct {
	AdminUserID adminspec.AdminUserID `json:"admin_user_id"`
}

func (r DisableUserRequest) Validate() []string {
	if !adminspec.IsAdminUserID(r.AdminUserID) {
		return []string{"admin_user_id"}
	}
	return []string{}
}

type EnableUserRequest struct {
	AdminUserID adminspec.AdminUserID `json:"admin_user_id"`
}

func (r EnableUserRequest) Validate() []string {
	if !adminspec.IsAdminUserID(r.AdminUserID) {
		return []string{"admin_user_id"}
	}
	return []string{}
}
